### DISCUSSION TITLE
Outlining issues with the Lazer editor in a hopefully constructive way

### DISCUSSION THREAD

**Discussion by Doksha** [ppy/osu#20706] [Ideas]
A lot of this stuff will be out of order, I will try to organize it after posting, I just want to get it out so I can get more opinions. This will basically just be a short essay on what is bad about the Lazer editor and what could be done differently ("your shit sucks" isn't good constructive criticism and won't get anything done). It is difficult to find a mapper who thinks the current editor is even mediocre, much less good. This is combining my personal opinion, and the opinions of other mappers.
.
.
Usually level editors for games are pretty difficult to get used to, but the osu stable editor was pretty quick to learn; the Lazer editor on the other hand is a pain in the ass. Ignoring the major bugs that need fixing, there are tons of things that make the Lazer editor more tedious to work with. The Lazer editor seems far too focused on simplicity over capability, with nearly every function from stable either being more difficult to execute, or entirely impossible.
.
.
###
### Issues with the comfort of editing a beatmap:

- When it comes to actual issues I and others have with the editor, I should start with object manipulation. None of the object manipulation options allow for text input, so rotating by any non 90 degree increments is slow, and scaling is quite difficult (additionally, scaling a selection of objects doesn’t move the points within a slider, although this is likely just a bug) I will talk about solutions to this in the hotkeys section. 
- Secondly, every time you select/manipulate an object, it makes a noise (this also applies to basically everything in Lazer), which can get irritating, there should be an option to disable the audio feedback in settings.
- There seems to be no way to edit sliderend sample and volume (maybe I am just missing something, but either way there needs to be an FAQ for stuff like this)
.
###
### Issues with UI:

- The timelines at the top and bottom of the screen are much larger (taller) than stable, as well as being opaque. This means much less of the screen is actually beatmap, and much more is useless space (the Lazer editor has the potential to actually have more stuff visible, just by making the menu buttons at the top of the screen thinner, the timelines shorter, and all options matching). My screen is 16:10 and even for me the timelines are way too intrusive. (I will cover this a little more later)
- On top of that, the real estate on the left and right (primarily the right) is barely used; all that needs to be done is to more closely mimic stable by having the button with the text of what that icon is (ie. the boxes on the left being select, circle, slider, and spinner, and the boxes on the right being combo, whistle, finish, clap, distance snap, and grid snap) this is already a good system, I don’t see any advantage to having them all on one side (and not showing the names of the object until you hover over them / click the hamburger)
- There is currently no way to dim the background, so working on a beatmap can become painful. 
- I understand the goal of trying to keep the dropdown menus minimal, but it is seriously holding Lazer back, there are so many simple things that could be added there that would improve the experience, such as resnapping, AImod, and revert to saved (open beatmap file would be nice, but I doubt that will be supported).
.
###
### Issues with workflow:

- Nobody likes the right click menu, the shift binding should be swapped (so that holding shift and right clicking brings up the menu, but normal right clicking would delete whatever hitobject / control point is under the cursor.) This menu could be made more streamlined if options were selected on mouse release, and not mouse click, this would allow for much better menu navigation. This doesn’t really apply to the other menus as they are basically useless anyway, but if they actually had options useful to the user, this would allow for more streamlined editing.
- Having options in entirely different tabs makes working in the editor clunky and tedious, the end result being more time spent in menus, and less time spent mapping.
- The timeline in the timing tab doesn’t even show hit circles, so if you easily forget what you are doing, you have to go back to the compose tab, check where you are, then return to timing (I don’t believe there would be any problems caused by showing the hitobjects on the timeline) It would be much more preferable if there was a way to bring up the timing tab as a popup menu instead of this entirely different menu.
.
###
### Hotkeys:

- Adding hotkeys for a few actions, such as changing the slider velocity (instead of having to click that tiny little green label above the slider), as well as the hit sounds, and most importantly, adding back the shift + number button combination for changing beat snap would make the Lazer editor much more usable.
- Hotkeys for rotate, scale, and translate would be quite useful (shift + r/g/s), when in the transform mode, inputting a numerical value would apply that value to the transform and tapping x/y in this mode would allow snapping to that axis (for translate and scale) 
.
###
I am going to sleep now, if I have any more ideas I will add them later; I just spent like 3 hours writing about how a video game’s level editor needs improvements. What is my life

**Comment by peppy** [comment]
Thanks for the feedback. A lot of what you've pointed out is already covered by other issues (and there are several issues like yours covering many areas). I'd recommend you read through them as I've already provided a lot of feedback/insight to the direction.

I can agree that there's still much potential to unlock. One initial suggestion would be to change your game UI scale if you find things too big. The toolbar elements in the editor take up less pixels compared to stable, so everything being bigger is a misconception.

**Reply by peppy** [reply]
> Having options in entirely different tabs makes working in the editor clunky and tedious, the end result being more time spent in menus, and less time spent mapping.

Not sure what this means.

> The timeline in the timing tab doesn’t even show hit circles

It didn't in stable either. The timeline may be removed on that screen going forward. It's intended only for timing. Actually changing SV etc. will be done from the main compose view.

> Usually level editors for games are pretty difficult to get used to, but the osu stable editor was pretty quick to learn; the Lazer editor on the other hand is a pain in the ass.

Keep in mind you're coming from the perspective of using osu! (stable) for years. That editor is less approachable to beginners with many hidden (but powerful) features, but does have some hint text which is missing from lazer (although we have better explanations in lazer for some things too)

**Reply by frenzibyte** [reply]
>> Having options in entirely different tabs makes working in the editor clunky and tedious, the end result being more time spent in menus, and less time spent mapping.
>
> Not sure what this means.

I'm going to guess this refers to the user being tired of switching between "compose" and "timing" screens, but the hotkeys should save time for that (F1 for compose, F2 for design, F3 for timing, F4 for setup). That said, a mapper should theoretically first set up the timing of the beatmap completely, and then switch to compose screen and start mapping.
